### PR TITLE
HTTP signed requests (client side)

### PR DESCRIPTION
**Pull request by jumaffre**
First step of https://github.com/microsoft/CCF/issues/546.

- Change `jsonrpc.py` (now `client.py`) to supports `requests`. `requests` let us use a custom authentication to sign the requests. We will use [requests-http-signature](https://github.com/pyauth/requests-http-signature/) for now until we hit its limitations (supported curves, etc.).
- Also refactored `client.py` as some code has become duplicated. There is now a generic `CCFClient` that has a custom client implementation (either framed TLS, requests or curl). For HTTP builds, we now default to `requests` but we still use `curl` when deploying a test network (by setting the `CURL_CLIENT` envvar). 
- Fixed a server-side issue that was only apparent when using `requests` (see inline comment).
- Also renamed `tests/client.py` to `test/perfclient.py` to avoid confusion.

### CODE REVIEW

File: src/enclave/http.h [hunk 1]
```diff
@@ -230,21 +229,25 @@ namespace enclave
 
       LOG_TRACE_FMT("recv called with {} bytes", size);
 
-      auto buf = read_all_available();
-
-      if (buf.size() == 0)
-        return;
+      while (true)
```
Comment (jumaffre) [inline src/enclave/http.h]: This is the fix that was required on the server side: for every `tls_inbound` message sent over the ring buffer, we read as much as 4k of decrypted TLS data from mbedtls and parse it.

**Comment by cimetrics** [comment]
images

File: samples/apps/txregulator/tests/txregulatorclient.py [hunk 2]
```diff
@@ -270,7 +270,7 @@ def run(args):
                             c.rpc("REG_get_revealed", {"tx_id": tx_id}),
                             error=lambda e: e is not None
                             and e["code"]
-                            == infra.jsonrpc.ErrorCode.INVALID_PARAMS.value,
+                            == infra.client.ErrorCode.INVALID_PARAMS.value,
```
Comment (eddyashton) [inline samples/apps/txregulator/tests/txregulatorclient.py]: I don't think these error codes should be moved into `infra.client` - they're JSON RPC specific, so ideally they'd stay under `infra.jsonrpc`.

File: tests/election.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 import math
 import infra.ccf
 import infra.proc
-import infra.jsonrpc
+import infra.client
```
Comment (eddyashton) [inline tests/election.py]: If the only change is renaming an import, then we're not actually using the import in this file - so many of these can simply be removed?

**Comment by cimetrics** [comment]
images

File: tests/election.py (hunk 3 above)
Comment (jumaffre) [reply tests/election.py]: You're absolutely right!

**Comment by cimetrics** [comment]
images

File: tests/infra/node.py [hunk 4]
```diff
@@ -188,38 +188,38 @@ def get_sealed_secrets(self):
         return self.remote.get_sealed_secrets()
 
     def user_client(self, format="msgpack", user_id=1, **kwargs):
-        return infra.jsonrpc.client(
+        return infra.client.client(
```
Comment (eddyashton) [inline tests/infra/node.py]: Nitpick: `infra.client.client` is a little awkward - `infra.clients`?

Comment (eddyashton) [reply tests/infra/node.py]: Or perhaps the function should be `make_client`

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: tests/infra/node.py (hunk 4 above)
Comment (jumaffre) [reply tests/infra/node.py]: I went for `infra.clients` as it supports different clients for now (requests, curl and framed TLS).

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

**Pull request merged**